src/extractvectors.1
.\" ==========================================================================
.\"         _   _      _   ____            __ __  __      _
.\"        | \ | | ___| |_|  _ \ ___ _ __ / _|  \/  | ___| |_ ___ _ __
.\"        |  \| |/ _ \ __| |_) / _ \ '__| |_| |\/| |/ _ \ __/ _ \ '__|
.\"        | |\  |  __/ |_|  __/  __/ |  |  _| |  | |  __/ ||  __/ |
.\"        |_| \_|\___|\__|_|   \___|_|  |_| |_|  |_|\___|\__\___|_|
.\"
.\"                  NetPerfMeter -- Network Performance Meter
.\"                 Copyright (C) 2009-2026 by Thomas Dreibholz
.\" ==========================================================================
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Contact:  dreibh@simula.no
.\" Homepage: https://www.nntb.no/~dreibh/netperfmeter/
.\"
.\" ###### Setup ############################################################
.Dd May 28, 2026
.Dt extractvectors 1
.Os ExtractVectors
.\" ###### Name #############################################################
.Sh NAME
.Nm extractvectors
.Nd Vector Extraction Tool for OMNeT++ Vector Files
.\" ###### Synopsis #########################################################
.Sh SYNOPSIS
.Nm extractvectors
.Op Ar input\_file
.Op Ar output\_file
.Op Ar [!]vector\_name\_prefix
.Op ...
.br
.Op Fl c Ar level | Fl \-compress Ar level
.br
.Op Fl s Ar separator | Fl \-separator Ar separator
.br
.Op Fl l | Fl \-line\-numbers | Fl n | Fl \-no\-line\-numbers
.br
.Op Fl p | Fl \-split | Fl a | Fl \-no\-split
.br
.Op Fl q | Fl \-quiet
.Nm extractvectors
.Op Fl h | Fl \-help
.Nm extractvectors
.Op Fl v | Fl \-version
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm extractvectors
is an extraction tool for vectors written by OMNeT++ to a data table. ExtractVectors supports on\-the\-fly BZip2 decompression and compression.
.Pp
.\" ###### Arguments ########################################################
.Sh ARGUMENTS
The following arguments may be provided:
.Bl -tag -width indent
.It Op Ar input\_file
The name of the input vector file.
.It Op Ar output\_file
The name of the output data table to be created.
.It Op Ar [!]vector\_name\_prefix
Name prefix of vectors to be extracted. Multiple vector name prefixes may be specified. If no name prefix is given, all vectors will be extracted. A "!" in front of the name turns vector splitting for this prefix on (see \-\-split option below).
.It Fl c Ar level | Fl \-compress Ar level
Sets the BZip2 compression level; 1=none, 9=highest (default).
.It Fl s Ar separator | Fl \-separator Ar separator
Sets the separator for the tables. Default: tabulator character (i.e. $'\t').
.It Fl l | Fl \-line\-numbers
Add line numbers into the output table files, as usual for GNU R.
.It Fl n | Fl \-no\-line\-numbers
Do not add line numbers into the output table files, as usual for CSV (default).
.It Fl p | Fl \-split
Try to split the vector name ending in a number into the name part as well as the number. A vector output table is then generated for the name part, with the corresponding number in table column "Split". Example: Vectors "Throughput 1", ..., "Throughput 9" will be written into output table as vector column "Throughput", the number will be in table column "Split".
\-\-split is a global option and applies to all vector name prefixes provided. To use vector name splitting for just a single prefix, prepend the prefix with "!".
.It Fl a | Fl \-no\-split
Turns vector splitting off (default).
.It Fl q | Fl \-quiet
Do not print verbose status information.
.It Fl h | Fl \-help
Prints command help.
.It Fl v | Fl \-version
Prints program version.
.El
.\" ###### Exit status ######################################################
.Sh EXIT STATUS
The
.Nm
tool exits with 0 on success, and >0 in case of an error.
.\" ###### Examples #########################################################
.Sh EXAMPLE
Consider an OMNeT++ vector file containing the vectors
"Slow Start Threshold <assoc>:<path>" and
"Congestion Window <assoc>:<path>",
contained with a lot of other vectors in the BZip2\-compressed vector file
.Pa results.vec.bz2 .
ExtractVectors can be used to extract all vectors beginning with "Slow Start Threshold" or "Congestion Window" into a BZip2\-compressed data
table
.Pa results.data.bz2
by the following command:
.br
extractvectors results.vec.bz2 results.data.bz2 \\
   "Slow Start Threshold" \\
   "Congestion Window"
.br
For each vector entry, an output line containing
time (Time), event number (Event),
object name (Object),
vector name (Vector),
name split parameter (Split), and
the vector value (Value) will be created.
.br
Example:
.br
Time     Event               Object Vector                          Split Value
.br
0.003672   174 "net.client[0].sctp" "Slow Start Threshold 2:Total"  ""    10000000
.br
0.003672   174 "net.client[0].sctp" "Slow Start Threshold 2:Total"  ""    10000000
.br
0.003672   174 "net.client[0].sctp" "Congestion Window 2:Total"     ""    4380
.br
0.003672   174 "net.client[0].sctp" "Congestion Window 2:Total"     ""    8760
.br
0.003913   238 "net.server[0].sctp" "Slow Start Threshold 4:Total"  ""    10000000
.br
0.003913   238 "net.server[0].sctp" "Slow Start Threshold 4:Total"  ""    10000000
.br
0.003913   238 "net.server[0].sctp" "Congestion Window 4:Total"     ""    3000
.br
0.003913   238 "net.server[0].sctp" "Congestion Window 4:Total"     ""    6000
.\" ###### See also #########################################################
.Sh SEE ALSO
.Xr extractvectors 1
.\" ###### Notes ############################################################
.Sh NOTES
This program is part of NetPerfMeter. The latest version of NetPerfMeter can be found on the NetPerfMeter Homepage at
.Lk https://www.nntb.no/\(tidreibh/netperfmeter/ "NetPerfMeter Homepage" .
.\" ###### Authors ##########################################################
.Sh AUTHORS
Thomas Dreibholz,
.Lk https://www.nntb.no/\(tidreibh "Homepage"


ChangeLog
commit 1f383a60d2d770535581e46f78444bcf9c4f4ebf
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 11:29:02 2026 +0200

    * Make inclusion of examples (scripts, results) optional in CMakeLists.txt.
    * Moved obsolete test program.

commit d26562086455bbf41d21f6c3cae99b6cae1bf2a4
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 10:43:28 2026 +0200

    Improved FreeBSD port.

commit 946bfafbadddcea276a7aa360c8ee697695c69db
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 10:42:01 2026 +0200

    Fixed includes.

commit 9d186d0746f498d32427611d54b5a927e5baa773
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 09:58:58 2026 +0200

    RPM spec fix.

commit 26e1ca4d015ec9403a22df90b8600605839e44e7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:13:55 2026 +0200

    New release netperfmeter-2.0.4.

commit 02576769d8562e11e6b0cb7d15200fe7d40e86cf
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:12:54 2026 +0200

    Minor manpage update.

commit bf272b25bfa5094a78b4657f4bd9912dcb23edb2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:03:05 2026 +0200

    Minor clean-up.

commit b732393412bd54a0e4d0f1347ede6747b7a7a870
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:01:02 2026 +0200

    Added option to configure the separator.

commit 977df2a565b3cf11fba2e83e8ad00f11c8c570ec
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:53:07 2026 +0200

    Small clean-ups.

commit 862fb32f3d7abcac2fff6537fba36b34981d9ffa
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:35:21 2026 +0200

    Further clean-ups.

commit 1d6d37a6dfac84a1858845f33fd9df66ff1d4806
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:31:15 2026 +0200

    Further clean-ups.

commit 793d87f6b1d0746c09488178c0e84424ce544115
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:30:39 2026 +0200

    Added option to configure the separator.

commit 2148af89133efdc3beeffc3f0c4a090e9cb75468
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 14:28:24 2026 +0200

    Improved shell completions.

commit 336be30bc58436d386d92183522a263cc30d1fd7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 14:22:41 2026 +0200

    Improved vector counting.

commit 8a5797d64ac46d71c275374869167aa5170cec12
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 14:19:01 2026 +0200

    Further improvements.

commit eb7c8be6269e2ca1790a0ec099201b3575516cc9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 13:33:00 2026 +0200

    Code improvements.

commit 05228c17745801c8910c7d06570f9ab804a20215
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 00:13:34 2026 +0200

    CMakeLists.txt update.

commit 1b427da57b4c7563f2418755c0439f4b8b7b076f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 00:08:58 2026 +0200

    Added more shell completions.

commit b799cc0179b7f960c9ef4d1fb28c5a9b34794e43
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:51:22 2026 +0200

    Options handling improvements.

commit 69b965905364e4306d1a8ef6149aa879136fca84
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:45:50 2026 +0200

    Packaging updates.

commit 82e1980494ab0fc8731b52671794c3c52a736c6b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:44:06 2026 +0200

    Added shell completions.

commit 8181abb986847144955c1e0b6e8da2c6dc863125
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:17:37 2026 +0200

    Further manpage improvements.

commit 3037cb9b5c9417375a288df233949c2e2ce9431a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:06:07 2026 +0200

    Improved manpages.

commit 428bdbaa9bccd79ed23d82f2a02157974ff70c51
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:05:57 2026 +0200

    Added version option.

commit 5eb240346cab65c110a07b63b0a2e4d748898b97
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:05:40 2026 +0200

    Added version option.

commit ef9d7f39317e1017354170936b4d5a359dce0f0e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 22:31:03 2026 +0200

    Improved CMakeLists.txt.

commit 58524abe9716d4f1a4f092f5aa64faec49fce966
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 21:03:35 2026 +0200

    Further manpage improvements.

commit 8c2c8ef1307836b855719897b2bcdb4775c11abc
Merge: 68c78ee b2d981b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 20:54:04 2026 +0200

    Merge branch 'master' of github.com:dreibh/netperfmeter

commit b2d981b677bc1bce0f2a9772a7905dff05375d3e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 16:46:28 2026 +0200

    Improved manpages.

commit dadcd3763d363a0a401d64439996f24eefe8780a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 25 22:37:19 2026 +0200

    A quick compile test under MacOS. Some more work would be needed to work around missing SCTP.

commit 68c78ee1c78dd0dc42e1131bae4064f039b1d7c3
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat May 9 10:36:06 2026 +0200

    Minor shell completion clean-up.

commit dcde13391c1f5dd16ce7f6db0da401713a3f310a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 5 15:20:48 2026 +0200

    New release netperfmeter-2.0.3.

commit 3ee0511e03d1e0ebf4f642c984ceb802f4345e9d
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 5 15:04:13 2026 +0200

    Another portability improvement.

commit ae4fc20b2ab2caec15c235b6d3f0a9348dffb6e5
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 5 15:03:18 2026 +0200

    Portability improvements.

commit 78f2b7ddc69f2a4385b8e9cfcde7dcf9f26471ae
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 5 14:56:40 2026 +0200

    Clean-up of the includes.

commit 1d415efe3466e5eb4dd1d7019031c34555c577ce
Author: Thomas Dreibholz <dreibh@simula.no>

src/extractvectors.cc
/*
 * ==========================================================================
 *         _   _      _   ____            __ __  __      _
 *        | \ | | ___| |_|  _ \ ___ _ __ / _|  \/  | ___| |_ ___ _ __
 *        |  \| |/ _ \ __| |_) / _ \ '__| |_| |\/| |/ _ \ __/ _ \ '__|
 *        | |\  |  __/ |_|  __/  __/ |  |  _| |  | |  __/ ||  __/ |
 *        |_| \_|\___|\__|_|   \___|_|  |_| |_|  |_|\___|\__\___|_|
 *
 *                  NetPerfMeter -- Network Performance Meter
 *                 Copyright (C) 2009-2026 by Thomas Dreibholz
 * ==========================================================================
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Contact:  dreibh@simula.no
 * Homepage: https://www.nntb.no/~dreibh/netperfmeter/
 */

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>
#include <iostream>
#include <map>
#include <string>
#include <unistd.h>
#include <vector>

#include "inputfile.h"
#include "outputfile.h"
#include "package-version.h"


class VectorInfo
{
   public:
   VectorInfo(const std::string& vectorPrefix, const bool splitMode) :
   VectorPrefix(vectorPrefix), SplitMode(splitMode), Found(false) { }
   std::string VectorPrefix;
   bool        SplitMode;
   bool        Found;
};


// ###### Read and process data file ########################################
static unsigned long long extractVectors(InputFile&               inputFile,
                                         OutputFile&              outputFile,
                                         const bool               defaultVectorSplittingMode,
                                         std::vector<VectorInfo>& vectorsToExtract,
                                         const bool               addLineNumbers,
                                         const char*              separator)
{
   std::map<unsigned int, const std::string> vectorToNameMap;
   std::map<unsigned int, const std::string> vectorToSplitMap;
   std::map<unsigned int, const std::string> vectorToObjectMap;
   unsigned long long                        outputLine = 0;
   char                                      inBuffer[4096];
   char                                      outBuffer[sizeof(inBuffer) + 4096];
   bool                                      versionOkay  = false;

   for(;;) {
      // ====== Read line from input file ===================================
      bool          eof;
      const ssize_t bytesRead = inputFile.readLine(inBuffer, sizeof(inBuffer), eof);
      if((bytesRead < 0) || (eof)) {
         break;
      }

      // ====== Process line ================================================
      outBuffer[0] = 0x00;
      if(!versionOkay) {
         // ====== Handle version number ====================================
         unsigned int version;
         if((inputFile.getLine() == 1) &&
            (sscanf(inBuffer, "version %u", &version) == 1)) {
            if(version != 2) {
               std::cerr << "ERROR: Got unknown version number "
                         << version << " (expected 2)\n";
               exit(1);
            }
            versionOkay = true;
            snprintf(outBuffer, sizeof(outBuffer),
                     "Time%sEvent%sObject%sVector%sSplit%sValue\n",
                     separator, separator, separator, separator, separator);
         }
         else {
            std::cerr << "ERROR: Missing \"version\" entry in input file!\n";
            exit(1);
         }
      }
      else {
         // ====== Handle data line =========================================
         unsigned int vectorID;
         unsigned int event;
         double       simTime;
         double       value;
         int          n;
         if(sscanf(inBuffer, "%u %u %lf %lf%n",
                   &vectorID, &event, &simTime, &value, &n) == 4) {
            std::map<unsigned int, const std::string>::iterator found =
               vectorToNameMap.find(vectorID);
            if(found != vectorToNameMap.end()) {
               if(addLineNumbers) {
                  snprintf(outBuffer, sizeof(outBuffer),
                           "%u%s%lf%s%u%s\"%s\"%s\"%s\"%s\"%s\"%s%lf\n",
                           (unsigned int)outputLine,            separator,
                           simTime,                             separator,
                           event,                               separator,
                           vectorToObjectMap[vectorID].c_str(), separator,
                           found->second.c_str(),               separator,
                           vectorToSplitMap[vectorID].c_str(),  separator,
                           value);
               }
               else {
                  snprintf(outBuffer, sizeof(outBuffer),
                           "%lf%s%u%s\"%s\"%s\"%s\"%s\"%s\"%s%lf\n",
                           simTime,                             separator,
                           event,                               separator,
                           vectorToObjectMap[vectorID].c_str(), separator,
                           found->second.c_str(),               separator,
                           vectorToSplitMap[vectorID].c_str(),  separator,
                           value);
               }
            }
         }

         // ====== Handle vector definition =================================
         else if(strncmp(inBuffer, "vector ", 7) == 0) {
            char objectName[1024];
            char vectorName[1024];
            if( (sscanf(inBuffer, "vector %u %1022s \"%1022[^\\\"]\" ETV",
                        &vectorID, objectName, vectorName) == 3) ||
                (sscanf(inBuffer, "vector %u %1022s %1022[^\\\"] ETV",
                        &vectorID, objectName, vectorName) == 3) ) {
               bool includeVector = (vectorsToExtract.size() == 0);
               bool splitMode     = defaultVectorSplittingMode;
               if(!includeVector) {
                  for(std::vector<VectorInfo>::iterator iterator = vectorsToExtract.begin();
                      iterator != vectorsToExtract.end(); iterator++) {
                     VectorInfo& vectorInfo = *iterator;
                     if(strncmp(vectorInfo.VectorPrefix.c_str(),
                                vectorName, vectorInfo.VectorPrefix.size()) == 0) {
                        includeVector = true;
                        vectorInfo.Found = true;
                        if(vectorInfo.SplitMode) {
                           splitMode = true;
                        }
                     }
                  }
               }

               if(includeVector) {
                  const char* splitName = "";
                  if(splitMode) {
                     for(int i = strlen(vectorName); i >= 0; i--) {
                        if(vectorName[i] == ' ') {
                           splitName     = &vectorName[i + 1];
                           vectorName[i] = 0x00;
                           break;
                        }
                     }
                     std::cout << "Adding vector \"" << vectorName << "\", split \""
                               << splitName << "\" of object " << objectName << " ...\n";
                  }
                  else {
                     std::cout << "Adding vector \"" << vectorName << "\" of object "
                               << objectName << " ...\n";
                  }

                  vectorToNameMap.insert(std::pair<unsigned int, const std::string>(
                     vectorID, std::string(vectorName)));
                  vectorToSplitMap.insert(std::pair<unsigned int, const std::string>(
                     vectorID, std::string(splitName)));
                  vectorToObjectMap.insert(std::pair<unsigned int, const std::string>(
                     vectorID, std::string(objectName)));
               }
            }
            else {
               std::cerr << "ERROR: Unexpected vector definition on input file "
                         << inputFile.getLine() << "!\n";
               exit(1);
            }
         }
      }


      // ====== Write output line ===========================================
      if(outBuffer[0] != 0x00) {
         if(!outputFile.write(outBuffer, strlen(outBuffer))) {
            exit(1);
         }
         outputLine++;
      }
   }

   // ====== Warn, if no vector had been extracted ==========================
   unsigned int uniqueFoundPrefixes = 0;
   for(std::vector<VectorInfo>::iterator iterator = vectorsToExtract.begin();
       iterator != vectorsToExtract.end(); iterator++) {
      if(iterator->Found) {
         uniqueFoundPrefixes++;
      }
   }
   if( (uniqueFoundPrefixes < vectorsToExtract.size()) &&
       (vectorsToExtract.size() != 0) ) {
      std::cerr << "WARNING: Found only " << uniqueFoundPrefixes << " of "
                << vectorsToExtract.size() << " specified!\n";
   }

   return outputLine;
}



// ###### Version ###########################################################
[[ noreturn ]] static void version()
{
   std::cerr << "ExtractVectors" << " " << EXTRACTVECTORS_VERSION << "\n";
   exit(0);
}


// ###### Usage #############################################################
[[ noreturn ]] static void usage(const char* program, const int exitCode)
{
   std::cerr << "Usage:\n"
      << "* Run:\n  "
      << program << "\n"
         "    input_file\n"
         "    output_file\n"
         "    [!]vector_name_prefix ...\n"
         "    [-c level|--compress level]\n"
         "    [-s separator|--separator separator]\n"
         "    [-l|--line-numbers|-n|--no-line-numbers]\n"
         "    [-p|--split|-a|--no-split]\n"
         "    [-q|--quiet]\n"
         "* Version:\n  " << program << " [-v|--version]\n"
         "* Help:\n  "    << program << " [-h|--help]\n";
   exit(exitCode);
}



// ###### Main program ######################################################
int main(int argc, char** argv)
{
   bool                    addLineNumbers      = false;
   unsigned int            compressionLevel    = 9;
   const char*             separator           = "\t";
   bool                    vectorSplittingMode = false;
   bool                    quietMode           = false;
   std::vector<VectorInfo> vectorsToExtract;


   // ====== Handle command-line arguments ==================================
   const static struct option long_options[] = {
      { "compress",        required_argument, 0, 'c' },
      { "separator",       required_argument, 0, 's' },
      { "line-numbers",    no_argument,       0, 'l' },
      { "no-line-numbers", no_argument,       0, 'n' },
      { "split",           no_argument,       0, 'p' },
      { "no-split",        no_argument,       0, 'a' },
      { "quiet",           no_argument,       0, 'q' },

      { "help",            no_argument,       0, 'h' },
      { "version",         no_argument,       0, 'v' },
      {  nullptr,          0,                 0, 0   }
   };

   int option;
   int longIndex;
   while( (option = getopt_long_only(argc, argv, "c:s:lnpaqhv", long_options, &longIndex)) != -1 ) {
      switch(option) {
         case 'c':
            compressionLevel = atol(optarg);
            if(compressionLevel < 1) {
               compressionLevel = 1;
            }
            else if(compressionLevel > 9) {
               compressionLevel = 9;
            }
          break;
         case 's':
            separator = optarg;
          break;
         case 'l':
            addLineNumbers = true;
          break;
         case 'n':
            addLineNumbers = false;
          break;
         case 'p':
            vectorSplittingMode = true;
          break;
         case 'a':
            vectorSplittingMode = false;
          break;
         case 'q':
            quietMode = true;
          break;
         case 'v':
            version();
          break;
         case 'h':
         case '?':
            // Exit with 0 on h/help, exit with 1 on '?' (unknown option):
            usage(argv[0], (option == 'h') ? 0 : 1);
          break;
         case '-':
          break;
         default:
            // This should not happen: wrong getopt parameters, or missing case?
            fprintf(stderr, "INTERNAL ERROR: Unhandled option c=%c code=%x!\n",
                    (isprint(option) ? (char)option : ' '), option);
            return 1;
          break;
      }
   }
   if(optind + 1 >= argc) {
      usage(argv[0], 1);
   }
   const std::string inputFileName(argv[optind++]);
   const std::string outputFileName(argv[optind++]);


   // ====== Print information ==============================================
   if(!quietMode) {
      std::cout << "ExtractVectors " << EXTRACTVECTORS_VERSION << "\n"
                << "* Vector Splitting:  " << (vectorSplittingMode  ? "on" : "off") << "\n"
                << "* Compression Level: " << compressionLevel << "\n"
                << "\n";
   }


   // ====== Open files =====================================================
   InputFile        inputFile;
   InputFileFormat  inputFileFormat = IFF_Plain;
   if( (inputFileName.size() >= 4) &&
       ( (inputFileName.substr(inputFileName.size() - 4) == ".bz2") ||
         (inputFileName.substr(inputFileName.size() - 4) == ".BZ2")) ) {
       inputFileFormat = IFF_BZip2;
   }
   if(inputFile.initialize(inputFileName.c_str(), inputFileFormat) == false) {
      exit(1);
   }

   OutputFile       outputFile;
   OutputFileFormat outputFileFormat = OFF_Plain;
   if( (outputFileName.size() >= 4) &&
       ( (outputFileName.substr(outputFileName.size() - 4) == ".bz2") ||
         (outputFileName.substr(outputFileName.size() - 4) == ".BZ2")) ) {
       outputFileFormat = OFF_BZip2;
   }
   if(outputFile.initialize(outputFileName.c_str(), outputFileFormat,
                            compressionLevel)== false) {
      exit(1);
   }


   // ====== Extract vectors ================================================
   for(int i = optind; i < argc; i++) {
      if(argv[i][0] == '!') {
         VectorInfo vectorInfo(&argv[i][1], true);
         vectorsToExtract.push_back(vectorInfo);
      }
      else {
         VectorInfo vectorInfo(argv[i], vectorSplittingMode);
         vectorsToExtract.push_back(vectorInfo);
      }
   }
   const unsigned long long lines =
      extractVectors(inputFile, outputFile,
                     vectorSplittingMode, vectorsToExtract,
                     addLineNumbers, separator);


   // ====== Close files ====================================================
   inputFile.finish();

   unsigned long long in, out;
   if(!outputFile.finish(true, &in, &out)) {
      exit(1);
   }
   if(!quietMode) {
      std::cout << "Wrote " << lines << " lines";
      if(in > 0) {
         std::cout << " (" << in << " -> " << out << " - "
                     << ((double)out * 100.0 / in) << "%)";
      }
      std::cout << "\n";
   }

   return 0;
}
